In [1]:
import jax
print(jax.__version__)

0.4.30


## Activation Functions

In [2]:
import numpy as np

In [3]:
"""
Sigmoid function
Sigmoid squashes input to a range between O and 1 , fitting well for binary classification tasks.

"""
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

In [4]:
"""
Rectified Linear Unit (ReLU)
ReLU is a non-linear activation function that outputs the input directly if it is positive, otherwise, it outputs zero.
Enhancing effciency compared to sigmoid.

"""
def relu(x):
    return np.maximum(0, x)

In [5]:
"""
Tanh function
Tanh squashes input to a range between -1 and 1, which is useful for classification tasks.
(different than Sigmoid that squashes input to a range between O and 1 )

"""
def tanh(x):
    return np.tanh(x)

In [6]:
"""
Softmax function
Softmax squashes input to a range between O and 1, and the sum of the output is 1.
Used in the output layer for multi-class classification, converting outputs into probabilities.

"""
def softmax(x):
    exps = np.exp(x - np.max(x, axis=1, keepdims=True))
    return exps / np.sum(exps, axis=1, keepdims=True)

## Backpropagation
Unraveling the Gradient Descent Algorithm.    
Backpropagation, the engine behind neural network training, iteratively adjusts weights and biases to minimize the
error between predicted and actual outputs.


In [7]:
# Assuming a simple neural network with one hidden layer
def backpropagation(inputs, targets, weights_input_hidden, weights_hidden_output):
  # Forward pass
  hidden_inputs = np.dot(inputs, weights_input_hidden)
  hidden_outputs = sigmoid(hidden_inputs)

  final_inputs = np.dot(hidden_outputs, weights_hidden_output)
  final_outputs = sigmoid(final_inputs)

  # Calculate error
  output_errors = targets - final_outputs

  # Backward pass
  output_grad = final_outputs * (1 - final_outputs) * output_errors
  hidden_errors = np.dot(output_grad, weights_hidden_output.T)
  hidden_grad = hidden_outputs * (1 - hidden_outputs) * hidden_errors

  # Update weights and biases
  weights_hidden_output += np.dot(hidden_outputs.T, output_grad)
  weights_input_hidden += np.dot(inputs.T, hidden_grad)

## Optimization techniques

### Stochastic Gradient Descent (SGD)
Stochastic Gradient Descent (SGD) is an optimization algorithm used to minimize a function by iteratively moving towards the minimum value of the function.     
It is particularly useful for training machine learning models.     
The primary goal of gradient descent is to identify the model parameters that provide the maximum accuracy on both training and test datasets.     
Unlike traditional Gradient Descent, which uses the entire dataset to compute the gradient of the loss function,    
SGD randomly selects a subset of data at each step.     
This makes SGD faster and more scalable for large datasets, though it may introduce more noise into the optimization process.

In [8]:
# Stochastic Gradient Descent SGD optimizer
def sgd_optimizer(inputs, targets, learning_rate=0.01, epochs=100):
  for epoch in range(epochs):
    for i in range(len(inputs)):
      # Forward pass
      # Backward pass and update weights
      print('.', end='')

### Momentum
A training optimization technique used to accelerate the convergence of gradient descent algorithms.    
It helps to speed up the learning process by incorporating the direction and velocity of the previous gradients into the current update.    
Essentially, it adds a fraction of the previous update step to the current update step,    
allowing the algorithm to move faster through shallow regions and dampening oscillations in steep regions.     
This technique is inspired by the physical concept of momentum, where an object in motion tends to stay in motion.    
In the context of neural network training, it helps to overcome some of the limitations of standard gradient descent    
by making the path towards the minimum more direct and thus potentially reducing the number of iterations needed to reach convergence.


In [9]:
def momentum_optimizer(inputs, targets, learning_rate=0.01, epochs=100, momentum=0.9):
  velocity = 0
  for epoch in range(epochs):
    for i in range(len(inputs)):
      # Forward pass
      # Backward pass and update weights
      velocity = momentum * velocity + learning_rate * gradient

### Adaptive Learning Rate
A training optimization technique that adjusts the learning rate dynamically during the training of a model.    
Unlike fixed learning rate strategies, adaptive methods modify the learning rate for each parameter,    
based on the history of gradients for that parameter.     
This approach helps in addressing issues like choosing an appropriate learning rate or adjusting it during training,     
which can significantly affect the convergence speed and quality of the final model.    
Techniques such as AdaGrad, RMSprop, and Adam are examples of adaptive learning rate methods.     
They aim to decrease the learning rate for parameters with large gradients to avoid overshooting and   
increase it for parameters with small gradients to speed up the learning process.    
This results in a more efficient and effective training process, especially for complex models and large datasets.

In [10]:
def adaptive_learning_rate_optimizer(inputs, targets, learning_rate=0.01, epochs=100):
  for epoch in range(epochs):
    for i in range(len(inputs)):
      # Forward pass
      # Backward pass and update weights
      learning_rate *= 1.0/(1.0 + learning_rate * epoch)

### Regularization Technique
A training optimization technique used to prevent overfitting in machine learning models.     
Overfitting occurs when a model learns the training data too well,     
capturing noise along with the underlying patterns, which results in poor performance on new, unseen data.     
Regularization addresses this issue by adding a penalty on the size of the model parameters to the loss function.     
This penalty encourages the model to keep the weights small, which can lead to simpler models that generalize better to new data.     
</br>
There are several types of regularization techniques, including:     
</br>
L1 Regularization (Lasso): Adds a penalty equal to the absolute value of the magnitude of coefficients.     
This can lead to sparse models where some weights can become zero, effectively performing feature selection.    
</br>
L2 Regularization (Ridge): Adds a penalty equal to the square of the magnitude of coefficients.     
This discourages large weights but does not necessarily drive them to zero.    
</br>
Elastic Net: Combines L1 and L2 regularization, adding both penalties to the loss function.    
This method enjoys the feature selection property of L1 and the smoothing of L2 regularization.   
</br>
Regularization techniques are widely used in linear regression, logistic regression, neural networks, and     
many other machine learning algorithms to improve their generalization capabilities.

In [11]:
def dropout(inputs, dropout_rate=0.2):
  mask = (np.random.rand(*inputs.shape) < 1.0 - dropout_rate) / (1.0 - dropout_rate)
  return inputs * mask

def weight_decay(weights, decay_rate=0.001):
  return weights - decay_rate * weights

## Implementing a Multilayer Perceptron (MLP)
Implement a simple multilayer perceptron (MLP) for a binary classification problem.    
Use NumPy for matrix operations and implement both forward and backward passes.     
Additionally, include a training loop to update the weights and biases using gradient descent.  </br>  

#### Requirements:
1. Design a multilayer perceptron with:
   - Input layer with 5 neurons.
   - Hidden layer with 10 neurons, using a ReLU activation function.
   - Output layer with 1 neuron and a sigmoid activation function.
2. Implement forward pass logic to compute the predicted output.
3. Implement backward pass logic to calculate gradients and update weights and biases using gradient descent.
4. Create a simple dataset for binary classification (e.g., use NumPy to generate random data).
5. Train your MLP on the dataset for a specified number of epochs.

<img src="../images/Developed-neural-network-of-5-10-1-for-the-present-application.png" alt="neural network" style="width:400px;"/>

<img src="../images/applsci-13-11329-g012.png" alt="neural network" style="width:600px;"/>

## Investigate parts in the MLP code

In [12]:
import numpy as np

# Define the MLP architecture
input_size = 5        # number of input layer neurons
hidden_size = 10      # number of hidden layer neurons
output_size = 1       # number of output layer neurons
learning_rate = 0.01

In [13]:
# Activation functions
def relu(x):
  return np.maximum(0, x)

def sigmoid(x):
  return 1 / (1 + np.exp(-x))

In [14]:
# weights and biases initialization
weights_input_hidden = np.random.randn(input_size, hidden_size)
biases_hidden = np.zeros((1, hidden_size))
# weights_hidden_output = np.random.randn(hidden_size, output_size)
# biases_output = np.zeros((1, output_size))
print(f"weights_input_hidden shape: {weights_input_hidden.shape}")
print(f"biases_hidden shape: {biases_hidden.shape}")
print(f"biases_hidden: {biases_hidden}")

weights_input_hidden shape: (5, 10)
biases_hidden shape: (1, 10)
biases_hidden: [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]


In [15]:
# Review the dataset
rows = 1 # I get an error in the backward pass investigate, if I try to pass a multi row dataset
         # [TODO] check ^^^....
np.random.seed(4)
X = np.random.rand(rows, input_size)
y = (X[:,0] + X[:,1] > 1).astype(int).reshape(-1, 1) # Reshape(-1, 1) - This operation will result in a 2D array with a shape (n, 1) , 
                                                     # where n is the number of elements in your original array. 
                                                     # This would be a column vector in this case

print(f"X: {X.shape}, y: {y.shape}")
print(f"X[:,0]: {X[:,0]}, \nX[:,1]: {X[:,1]},\n>1: {(X[:,0] + X[:,1] > 1)}, \ny: {y.T}") # y.T: transposing the y column vector into a row vector
                                                                                         # for nice printing
for i in range(rows):
  print(f"{i}) X: {X[i]}, y: {y[i]}")

print(f"X[0]: {X[0]}")

X: (1, 5), y: (1, 1)
X[:,0]: [0.96702984], 
X[:,1]: [0.54723225],
>1: [ True], 
y: [[1]]
0) X: [0.96702984 0.54723225 0.97268436 0.71481599 0.69772882], y: [1]
X[0]: [0.96702984 0.54723225 0.97268436 0.71481599 0.69772882]


In [16]:
# Calculate the test sample
data_point = np.array([[0.6, 0.7, 0.8, 0.9, 1.0]])
target = (data_point[0,0] + data_point[0,1] > 1).astype(int)
print(f"data_point[0,0]: {data_point[0,0]} + data_point[0,1]: {data_point[0,1]}, target: {target}")

data_point = np.array([[0.1, 0.2, 0.3, 0.4, 0.5]])
target = (data_point[0,0] + data_point[0,1] > 1).astype(int)
print(f"data_point[0,0]: {data_point[0,0]} + data_point[0,1]: {data_point[0,1]}, target: {target}")

data_point[0,0]: 0.6 + data_point[0,1]: 0.7, target: 1
data_point[0,0]: 0.1 + data_point[0,1]: 0.2, target: 0


In [17]:
# def forward_pass(inputs):
#   global weights_hidden_output, biases_output, weights_input_hidden, biases_hidden
#   hidden_input = np.dot(inputs, weights_input_hidden) + biases_hidden
#   hidden_output = relu(hidden_input)
#   final_input = np.dot(hidden_output, weights_hidden_output) + biases_output
#   predicted_output = sigmoid(final_input)
#   return predicted_output, hidden_output


inputs = np.array([[0.6, 0.7, 0.8, 0.9, 1.0]]) # data input array
weights_input_hidden = np.random.randn(input_size, hidden_size) # weights connecting the input to the hidden layer
biases_hidden = np.zeros((1, hidden_size)) # biases for the hidden layer
weights_hidden_output = np.random.randn(hidden_size, output_size)
biases_output = np.zeros((1, output_size))

print(f"weights_hidden_output: {weights_hidden_output}")

# inputs to the neurons in the hidden layer
hidden_input = np.dot(inputs, weights_input_hidden) + biases_hidden
print(f"\ninputs to the neurons in the hidden layer: \n{hidden_input}")

hidden_output = relu(hidden_input) # relu: input directly if it is positive, otherwise, it outputs zero.
print(f"\noutput of the neurons in the hidden layer after relu activation function: \n{hidden_output}")

# inputs to the neurons in the final layer
final_input = np.dot(hidden_output, weights_hidden_output) + biases_output
print(f"\ninputs to the neurons in the final layer: \n{final_input}")

predicted_output = sigmoid(final_input) # sigmoid: squashes input to a range between O and 1
print(f"\noutput of the neurons in the final layer after sigmoid activation function: \n{predicted_output}")

weights_hidden_output: [[ 0.82242392]
 [-0.90004667]
 [ 0.0362415 ]
 [ 0.97073319]
 [-0.65066811]
 [ 0.15493329]
 [-1.0765125 ]
 [ 0.11634145]
 [-1.55579813]
 [ 1.3072029 ]]

inputs to the neurons in the hidden layer: 
[[ 1.19900095 -0.97887772 -1.27884885 -0.10244675  3.68446011 -2.29960953
  -0.10745545 -0.27106168  3.77135906  0.99667761]]

output of the neurons in the hidden layer after relu activation function: 
[[1.19900095 0.         0.         0.         3.68446011 0.
  0.         0.         3.77135906 0.99667761]]

inputs to the neurons in the final layer: 
[[-5.97588713]]

output of the neurons in the final layer after sigmoid activation function: 
[[0.00253282]]


In [18]:
# # Backward pass
# def backward_pass(inputs, predicted_output, hidden_output, targets):
#   global weights_hidden_output, biases_output, weights_input_hidden, biases_hidden
#   output_error = predicted_output - targets
#   output_delta = output_error * (predicted_output * (1 - predicted_output))

#   hidden_error = output_delta.dot(weights_hidden_output.T)
#   hidden_delta = hidden_error * (hidden_output > 0)

#   # Update weights and biases
#   weights_hidden_output -= learning_rate * hidden_output.T.dot(output_delta)
#   biases_output -= learning_rate * np.sum(output_delta, axis=0, keepdims=True)
#   weights_input_hidden -= learning_rate * inputs.T.dot(hidden_delta)
#   biases_hidden -= learning_rate * np.sum(hidden_delta, axis=0, keepdims=True)

# backward_pass(X, predicted_output, hidden_output, y)
print(f"X: {X} \ny: {y.T} \npredicted_output: {predicted_output} \nhidden_output: {hidden_output}")

inputs = X
targets = y
output_error = predicted_output[0] - targets
print(f"\noutput_error: \n{output_error}")

output_delta = output_error * (predicted_output * (1 - predicted_output))
print(f"\noutput_delta: \n{output_delta}")

hidden_error = output_delta.dot(weights_hidden_output.T)
print(f"\nhidden_error: \n{hidden_error}")

hidden_delta = hidden_error * (hidden_output[0] > 0)
print(f"\nhidden_delta: \n{hidden_delta}")

weights_hidden_output -= learning_rate * hidden_output.T.dot(output_delta)
print(f"\nweights_hidden_output: \n{weights_hidden_output}")

biases_output -= learning_rate * np.sum(output_delta, axis=0, keepdims=True)
# print(f"\nbiases_output: \n{biases_output}")

weights_input_hidden -= learning_rate * inputs.T.dot(hidden_delta)
# print(f"\nweights_input_hidden: \n{weights_input_hidden}")

biases_hidden -= learning_rate * np.sum(hidden_delta, axis=0, keepdims=True)
#print(f"\nbiases_hidden: \n{biases_hidden}")

X: [[0.96702984 0.54723225 0.97268436 0.71481599 0.69772882]] 
y: [[1]] 
predicted_output: [[0.00253282]] 
hidden_output: [[1.19900095 0.         0.         0.         3.68446011 0.
  0.         0.         3.77135906 0.99667761]]

output_error: 
[[-0.99746718]]

output_delta: 
[[-0.00252]]

hidden_error: 
[[-2.07251068e-03  2.26812024e-03 -9.13286845e-05 -2.44625048e-03
   1.63968554e-03 -3.90432336e-04  2.71281465e-03 -2.93180795e-04
   3.92061582e-03 -3.29415513e-03]]

hidden_delta: 
[[-0.00207251  0.         -0.         -0.          0.00163969 -0.
   0.         -0.          0.00392062 -0.00329416]]

weights_hidden_output: 
[[ 0.82245414]
 [-0.90004667]
 [ 0.0362415 ]
 [ 0.97073319]
 [-0.65057526]
 [ 0.15493329]
 [-1.0765125 ]
 [ 0.11634145]
 [-1.55570309]
 [ 1.30722802]]


#### Comparing the weights_hidden_output before and after the backward pass

    Before           After                
 </br>
[[ 0.82242392] ---  [[ 0.82245414]  </br>        
 [-0.90004667] ---   [-0.90004667]  </br>     
 [ 0.0362415 ] ---   [ 0.0362415 ]  </br>          
 [ 0.97073319] ---   [ 0.97073319]  </br>
 [-0.65066811] ---   [-0.65057526]  </br>
 [ 0.15493329] ---   [ 0.15493329]  </br>
 [-1.0765125 ] ---   [-1.0765125 ]  </br>
 [ 0.11634145] ---   [ 0.11634145]  </br>
 [-1.55579813] ---   [-1.55570309]  </br>
 [ 1.3072029 ]]---   [ 1.30722802]] </br>

 </br>
 </br>

## MLP implementation

In [19]:
# Multilayer Perceptron (MLP)
import numpy as np

# Define the MLP architecture
input_size = 5        # number of input layer neurons
hidden_size = 10      # number of hidden layer neurons
output_size = 1       # number of output layer neurons
learning_rate = 0.01  # a hyperparameter determines the amount that the weights are updated during training
epochs = 1000

# Initialize weights and biases
weights_input_hidden = np.random.randn(input_size, hidden_size)
biases_hidden = np.zeros((1, hidden_size))
weights_hidden_output = np.random.randn(hidden_size, output_size)
biases_output = np.zeros((1, output_size))

# Activation functions
def relu(x):
  return np.maximum(0, x)

def sigmoid(x):
  return 1 / (1 + np.exp(-x))

# Forward pass
def forward_pass(inputs):
  global weights_hidden_output, biases_output, weights_input_hidden, biases_hidden
  hidden_input = np.dot(inputs, weights_input_hidden) + biases_hidden
  hidden_output = relu(hidden_input)
  final_input = np.dot(hidden_output, weights_hidden_output) + biases_output
  predicted_output = sigmoid(final_input)
  return predicted_output, hidden_output

# Backward pass
def backward_pass(inputs, predicted_output, hidden_output, targets):
  global weights_hidden_output, biases_output, weights_input_hidden, biases_hidden
  output_error = predicted_output - targets
  output_delta = output_error * (predicted_output * (1 - predicted_output))

  hidden_error = output_delta.dot(weights_hidden_output.T)
  hidden_delta = hidden_error * (hidden_output > 0)

  # Update weights and biases
  weights_hidden_output -= learning_rate * hidden_output.T.dot(output_delta)
  biases_output -= learning_rate * np.sum(output_delta, axis=0, keepdims=True)
  weights_input_hidden -= learning_rate * inputs.T.dot(hidden_delta)
  biases_hidden -= learning_rate * np.sum(hidden_delta, axis=0, keepdims=True)

# Generate a simple dataset
np.random.seed(0)
X = np.random.rand(100, input_size)
y = (X[:,0] + X[:,1] > 1).astype(int).reshape(-1, 1)

# Training loop
for epoch in range(epochs):
  # Forward pass
  predicted_output, hidden_output = forward_pass(X)

  # Backward pass
  backward_pass(X, predicted_output, hidden_output, y)

  # Print loss every 100 epochs
  if epoch % 100 == 0:
    loss = -np.mean(y * np.log(predicted_output) + (1 - y) * np.log(1 - predicted_output))
    print(f"Epoch {epoch}, loss {loss}")

# Test the model
data_point = np.array([[0.6, 0.7, 0.8, 0.9, 1.0]])
predicted_output, _ = forward_pass(data_point)
print(f"Predicted output for data point: {predicted_output}")

data_point = np.array([[0.1, 0.2, 0.3, 0.4, 0.5]])
predicted_output, _ = forward_pass(data_point)
print(f"Predicted output for data point: {predicted_output}")

Epoch 0, loss 0.9918718317468016
Epoch 100, loss 0.323283828330311
Epoch 200, loss 0.21002507455021519
Epoch 300, loss 0.16220571186563126
Epoch 400, loss 0.1349856200943552
Epoch 500, loss 0.11718358369016474
Epoch 600, loss 0.10408923335581735
Epoch 700, loss 0.09382738965757148
Epoch 800, loss 0.08562899267325655
Epoch 900, loss 0.07894051787603523
Predicted output for data point: [[0.98943064]]
Predicted output for data point: [[0.00385779]]
